Journalist can create and modify article

piavW opened the pull request:
[PT-link](https://www.pivotaltracker.com/story/show/169282741)

## PR Content
Creates functionality for an editor to give a journalist access to create an article (makes it possible to create an article). 

## This PR Contains
- Unit tests of user model
- Pundit gem
- Create action for article model 
- Possibility to attach image to article
- Roles attributes to user model
- Create & edit action for user model
- Request tests for article model create, update, index actions
- Request test for authorization of create and update actions

## What we've learned
- How to use pundit and pundit-matchers
- How to set a policy and test that policy
- How to create associations with a role of a user model

piavW committed:
Creates test-file, adds post test and create article test.

yasmineezequiel committed:
Adds create action with image attached action for article.

beccaburns committed:
Creates services module for image attachment. Updates spec fil and controller. Adds has_one_attached association for image in article model.

piavW committed:
Configures roles for user model
- installs Pundit for authorization of users
- migrates role into database table
- adds users with different roles to seed
- adds articles to seed
- updates tests to include role in user model

yasmineezequiel committed:
Installed pundit-matchers, defines roles for user.

beccaburns committed:
Configures policies for article
- Adds Application policy
- Adds Article policy
- Incorporates Pundit
- Removes unnecessary pundit matcher user alias config in rails helper
- Creates spec test for article policies

piavW committed:
Edit functionality of article
- adds spec-test for edit functionality
- updates articles policy regarding editing
- updates articles policy test
- adds actions to articles controller

yasmineezequiel committed:
Adds association between user and article, updates test.

yasmineezequiel committed:
Modifies article create spec and moves devise token mount out from the api name space

Co-authored-by: Thomas Ochman <thomas@craftacademy.se>

yasmineezequiel committed:
Adds sad path  test for user that is not a journalist returning 404 error

Co-authored-by: Thomas Ochman <thomas@craftacademy.se>

yasmineezequiel committed:
Adds pundit in Application controller and defines user not authorized,  test error reponse in spec.

Co-author Thomas Ochman

piavW committed:
Adds association to article factory.

beccaburns committed:
Updates policy spec for subscriber & journalist.

DIzosimov commented on app/controllers/api/v1/articles_controller.rb:
```diff
@@ -1,4 +1,5 @@
 class Api::V1::ArticlesController < ApplicationController
+  before_action :authenticate_user!
```
space between before action and index

yasmineezequiel committed:
[WIP] working on solving tests after pundit authorization implementation.

faraznaeem committed:
Adds user to specs

piavW committed:
Installs Faker gem, updates Factory user with faker email and nickname.

piavW committed:
Refactors request-tests and articles controller.

DIzosimov commented on spec/models/article_spec.rb:
```diff
@@ -16,7 +16,11 @@
   describe 'Factory' do
     it 'should have valid Factory' do
-      expect(create(:article)).to be_valid 
+      expect( create(:article)).to be_valid 
```
random space before create

DIzosimov reviewed (changes requested):
check comment (random space thats need to be removed)

piavW committed:
Adds new line and removes space.

faraznaeem commented on config/routes.rb:
```diff
@@ -1,12 +1,14 @@
 
     namespace :v1 do
-      mount_devise_token_auth_for 'User', at: 'auth', skip: [:omniauth_callbacks]
+      # mount_devise_token_auth_for 'User', at: 'auth', skip: [:omniauth_callbacks]
```
Please clean out the comments

faraznaeem commented on spec/models/article_spec.rb:
```diff
@@ -19,4 +19,8 @@
       expect(create(:article)).to be_valid 
     end
   end
+  
```
Add a test for the attached image

yasmineezequiel committed:
Adds image Factory validation and test.

yasmineezequiel committed:
Eliminates comments from routes.

tochman commented on app/controllers/api/v1/articles_controller.rb:
```diff
@@ -9,9 +11,48 @@ def index
+  end
+
+  def update
+    @article = Article.find(params[:id])
```
You need to authorize the user for `update` action

tochman commented on app/policies/article_policy.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  def new?
+    @user.journalist?
```
There is no `new` actions in API's

tochman commented on spec/models/article_spec.rb:
```diff
@@ -19,4 +19,19 @@
   end
+  
+  it 'Should have a journalist' do
+    should belong_to(:journalist)
```
?? 😳

piavW replied:
Yea that test isn't doing anything really is it...

tochman commented on spec/policies/article_policy_spec.rb:
```diff
@@ -0,0 +1,34 @@
+
+describe ArticlePolicy do
+  subject { described_class.new(user, article) }
+  let(:user) { User.create(role: 'journalist') }
```
Please use factories when creating objects in tests

faraznaeem replied:
Please address this

piavW replied:
my bad, missed that one when I changed the other ones.

tochman reviewed (changes requested):
There are issues on this one

piavW committed:
Refactors tests, attempts to requre authentization of user for edit-action.

piavW committed:
Refactors authorization of update action.

piavW committed:
Refactors tests with sad paths
- Adds it-block with sad-path error message & status for edit article
- Updates create-article with sadpath error message

piavW committed:
Removes .DS_Store

piavW committed:
Refactors code to include factorybot usage.
